Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py

```diff
@@ -83,27 +87,67 @@ def editable(cls, branch, repository=None):
             return False
 
         remote_repo = repository.remote(name=repository.default_remote)
-        if remote_repo and isinstance(remote_repo, remote.GitHub):
-            for name in repository.source_remotes():
-                if branch in repository.branches_for(remote=name, cached=True):
-                    return False
+        if remote and remote_repo and isinstance(remote_repo, remote.GitHub):
+            return remote in repository.source_remotes()
         return True
 
     @classmethod
     def branch_point(cls, repository, limit=None):
         cnt = 0
-        commit = repository.commit(include_log=False, include_identifier=False)
-        while cls.editable(commit.branch, repository=repository):
-            cnt += 1
-            if limit and cnt > limit:
+        try:
+            proc = subprocess.Popen(
+                [repository.executable(), 'rev-list', 'HEAD'],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                cwd=repository.root_path,
+                **(dict(encoding='utf-8') if sys.version_info >= (3, 6) else dict())
+            )
+            if proc.poll():
                 return None
-            commit = repository.find(argument='HEAD~{}'.format(cnt), include_log=False, include_identifier=False)
-            if cnt > 1 or commit.branch != repository.branch or cls.editable(commit.branch, repository=repository):
-                log.info('    Found {}...'.format(string_utils.pluralize(cnt, 'commit')))
-            else:
-                log.info('    No commits on editable branch')
+            commit_hash = proc.stdout.readline().strip()
+            while commit_hash:
+                if limit and cnt > limit:
+                    return None
+
+                # If the specified commit is on the default branch, we've definitively found the branch point
+                if not run(
+                    [repository.executable(), 'merge-base', '--is-ancestor', commit_hash, repository.default_branch],
+                    cwd=repository.root_path,
+                    capture_output=True,
+                ).returncode:
+                    break
 
-        return commit
+                # Otherwise, check if any of the branches this commit is on are not editable
+                for branch_line in run(
+                    [repository.executable(), 'branch', '--contains', commit_hash, '-a', '--format', '%(refname)'],
+                    cwd=repository.root_path,
+                    capture_output=True,
+                    encoding='utf-8',
+                ).stdout.splitlines():
```

Comment:
Given how expensive this is (taking many seconds per call, as it has to walk the entire DAG for each call), could we instead look at the commit message and pull out the branch from the canonical link? Because that branch is likely to contain this commit and is likely to not be editable.

We'll still need to fall back to this, but in the common case we can avoid the expensive call.

Reply:
I think falling back isn't going to help, because you've got the common case backwards. The common case is actually that the commit we're referencing isn't on a production branch, it's on a PR branch. In that case (as you point out) we have to fall back to the more expensive call anyways.

It's also worth pointing out that this call is as expensive as the number of branches which contain a commit. That will be most dramatic if the commit is on `main`. Except we won't run this code if the commit is on `main` (because the `merge-base` call will have caught that case).